pkg/server/secrets-encrypt.go

```diff
@@ -0,0 +1,508 @@
+			default:
+				err = fmt.Errorf("unknown stage %s requested", *encryptReq.Stage)
+			}
+		} else if encryptReq.Enable != nil {
+			err = encryptionEnable(ctx, server, *encryptReq.Enable)
+		}
+
+		if err != nil {
+			genErrorMessage(resp, http.StatusBadRequest, err)
+			return
+		}
+		resp.WriteHeader(http.StatusOK)
+	})
+}
+
+func encryptionPrepare(ctx context.Context, server *config.Control, force bool) error {
+	state, err := getEncryptionState(server)
+	if err != nil {
+		return err
+	} else if !force && (state.Stage != EncryptionStart && state.Stage != EncryptionReencrypt) {
+		return fmt.Errorf("error, incorrect stage: %s found with key: %s", state.Stage, state.ActiveKey)
+	}
+
+	if err := verifyEncryptionHashAnnotation(server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+
+	curKeys, err := getEncryptionKeys(server)
+	if err != nil {
+		return err
+	}
+
+	if err := AppendNewEncryptionKey(&curKeys); err != nil {
+		return err
+	}
+	logrus.Infoln("Adding secrets-encryption key: ", curKeys[len(curKeys)-1])
+
+	if err := writeEncryptionConfig(server, curKeys, true); err != nil {
+		return err
+	}
+	if err := writeEncryptionState(server, EncryptionPrepare, curKeys[0]); err != nil {
+		return err
+	}
+	if err := writeEncryptionHashAnnotation(server, server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+	return cluster.Save(ctx, server, server.Runtime.EtcdConfig, true)
+}
+
+func encryptionRotate(ctx context.Context, server *config.Control, force bool) error {
+	state, err := getEncryptionState(server)
+	if err != nil {
+		return err
+	} else if !force && state.Stage != EncryptionPrepare {
+		return fmt.Errorf("error, incorrect stage: %s found with key: %s", state.Stage, state.ActiveKey)
+	}
+
+	if err := verifyEncryptionHashAnnotation(server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+
+	curKeys, err := getEncryptionKeys(server)
+	if err != nil {
+		return err
+	}
+
+	// Right rotate elements
+	rotatedKeys := append(curKeys[len(curKeys)-1:], curKeys[:len(curKeys)-1]...)
+
+	if err = writeEncryptionConfig(server, rotatedKeys, true); err != nil {
+		return err
+	}
+	if err := writeEncryptionState(server, EncryptionRotate, curKeys[0]); err != nil {
+		return err
+	}
+	logrus.Infoln("Encryption keys right rotated")
+	if err := writeEncryptionHashAnnotation(server, server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+	return cluster.Save(ctx, server, server.Runtime.EtcdConfig, true)
+}
+
+func encryptionReencrypt(ctx context.Context, server *config.Control, force bool, skip bool) error {
+
+	state, err := getEncryptionState(server)
+	if err != nil {
+		return err
+	} else if !force && state.Stage != EncryptionRotate {
+		return fmt.Errorf("error, incorrect stage: %s found with key: %s", state.Stage, state.ActiveKey)
+	}
+	if err := verifyEncryptionHashAnnotation(server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+
+	updateSecrets(server.Runtime.Core.Core())
```

Comment:
You might talk with @Oats87 for potential concerns. Keep in mind that some clusters may have thousands of secrets and this may take more than a few hours to complete.

I suspect you're going to want to make this a goroutine, and probably have some way to track its progress (possibly via `v1.Event` records, or a status field somewhere?) and whether or not it's completed successfully.

Reply:
Responded to error if updatingSecrets fails, and bail out before we remove the old key. Also include in the error how much progress was made (X out of Y secrets reencrypted). This state is still okay for the server, we still have both keys, so the user can simply rerun `reencrypt` with no consequence, just some key will be "reencrypted" with the new key twice.

Reply:
Converted reencrypt to a controller for secrets encryption